### PR TITLE
Create version.txt to list package versions on servers

### PR DESCRIPTION
**Pull request by karenc**
This creates `/var/www/webview/version.txt` on all the frontend servers
with versions of webview, archive, publishing, zope packages and the
date when version.txt was updated.  Future versions will be prepended to
version.txt (things on the top are most recent) so we get a list of what
version of packages were on the servers when.

An example version.txt that has been deployed once looks like this:

```
2017-10-06 14:08:39 PDT

webview==50042368add58b0cf4889ac5e37e8a55d09e1ef6

cnx-archive==2.6.1

cnx-publishing==0.9.3

-------------------------------
Archive full versions from staging07.cnx.org:
appdirs==1.4.3
Beaker==1.8.0
cnx-archive==2.6.1
cnx-db==0.10.4
cnx-epub==0.12.0
cnx-query-grammar==0.2.2
db-migrator==1.0.1
funcsigs==1.0.2
Jinja2==2.7.3
lxml==3.4.0
MarkupSafe==0.23
packaging==16.8
parsimonious==0.6.1
Paste==1.7.5.1
PasteDeploy==1.5.2
PasteScript==1.7.5
pkg-resources==0.0.0
psycopg2==2.7.3.1
pyparsing==2.2.0
pyramid==1.6a2
pyramid-jinja2==2.6.2
pyramid-multiauth==0.8.0
pyramid-sawing==1.1.2
pyramid-translogger==0.1
python-memcached==1.53
pytz==2017.2
PyYAML==3.12
repoze.lru==0.6
requests==2.7.0
rhaptos.cnxmlutils==1.3.0
sanction==0.4.1
six==1.10.0
translationstring==1.3
tzlocal==1.1.3
venusian==1.0
waitress==1.0.2
webencodings==0.5
WebOb==1.5.0a1
zope.deprecation==4.1.2
zope.interface==4.1.2

-------------------------------
Publishing full versions from staging08.cnx.org:
amqp==2.2.2
appdirs==1.4.3
Beaker==1.8.0
billiard==3.5.0.3
celery==4.1.0
cnx-archive==2.6.1
cnx-db==0.10.4
cnx-easybake==0.7.0
cnx-epub==0.12.0
cnx-publishing==0.9.3
cnx-query-grammar==0.2.2
cssselect==0.9.1
-e git+https://github.com/Connexions/cssselect2.git@560b4c432fcea05e4694ad6367b5a0e4d9fe2d33#egg=cssselect2
db-migrator==1.0.1
funcsigs==1.0.2
Jinja2==2.7.3
kombu==4.1.0
lxml==3.4.0
MarkupSafe==0.23
openstax-accounts==1.0.0
packaging==16.8
parsimonious==0.6.1
Paste==1.7.5.1
PasteDeploy==1.5.2
PasteScript==1.7.5
pkg-resources==0.0.0
psycopg2==2.7.3.1
pyparsing==2.2.0
pyramid==1.6a2
pyramid-jinja2==2.6.2
pyramid-multiauth==0.8.0
pyramid-sawing==1.1.2
pyramid-translogger==0.1
python-memcached==1.53
pytz==2017.2
PyYAML==3.12
repoze.lru==0.6
requests==2.7.0
rhaptos.cnxmlutils==1.3.0
sanction==0.4.1
six==1.10.0
SQLAlchemy==1.1.14
tinycss2==0.5
translationstring==1.3
tzlocal==1.1.3
venusian==1.0
vine==1.1.4
waitress==1.0.2
webencodings==0.5
WebOb==1.5.0a1
zope.deprecation==4.1.2
zope.interface==4.1.2

-------------------------------
Legacy full versions from staging04.cnx.org:
Jinja2==2.6
Pillow==1.7.6
Products.AdvancedQuery==2.2-rhaptosdev-r30378
Products.CMFDiffTool==0.3.3dev-r30461
Products.CMFSquidTool==1.5.1
Products.CNXCaching==0.2
Products.CNXContent==0.47
Products.CNXMLDocument==0.49
Products.CNXMLTransforms==0.23
Products.CNXPloneSite==0.76.2dev
Products.CacheSetup==1.2.1
Products.CatalogMemberDataTool==1.5
Products.ClockServer==0.2-Zope2.9dev
Products.ExtZSQL==0.9
Products.ExternalFile==0.1-rhaptosdev-r30470
Products.ExternalStorage==0.8.1
Products.FSImportTool==0.8
Products.FeatureArticle==0.10.1
Products.Five==1.4.5-rhaptosdev-r32037
Products.LensOrganizer==0.3.1
Products.Lensmaker==0.25
Products.Lineup==0.6
Products.LinkMapTool==0.9
Products.LocalFS==1.7-rhaptosdev-r30378
Products.ManagableIndex==1.7.3-rhaptosdev-r30378
Products.MasterSelectWidget==0.2.3-rhaptosdev-r30378
Products.MathEditor==0.4.1
Products.NoHeaderFieldContinuation==0.1-rhaptosdev-r30378
Products.OFolder==1.0-rhaptosdev-r30378
Products.PageCacheManager==1.2
Products.PloneHotfix20110531==2.0
Products.PloneHotfix20110720==1.2
Products.PloneHotfix20121106==1.2
Products.PloneHotfix20130618==1.1
Products.PolicyHTTPCacheManager==1.2
Products.References==0.10-rhaptosdev-r30378
Products.RhaptosBugTrackingTool==0.1.3
Products.RhaptosCacheTool==0.2.1
Products.RhaptosCatFixes==0.2
Products.RhaptosCollaborationTool==0.10.1
Products.RhaptosCollection==0.77
Products.RhaptosContent==0.66
Products.RhaptosHitCountTool==0.12
Products.RhaptosModuleEditor==0.84
Products.RhaptosModuleStorage==1.1.3
Products.RhaptosPDFLatexTool==0.15.1dev
Products.RhaptosPatchTool==0.7.2
Products.RhaptosPrint==0.104
Products.RhaptosRepository==1.4.2
Products.RhaptosSimilarityTool==0.13
Products.RhaptosSite==1.46
Products.RhaptosSword==2.3.2dev
Products.RhaptosTest==1.0
Products.RhaptosWorkgroup==0.27
Products.RisaCollection==0.1
Products.RisaHitCountTool==0.1
Products.RisaModuleEditor==0.1
Products.RisaModuleStorage==0.1
Products.RisaPDFLatexTool==0.1
Products.RisaPatchTool==0.1
Products.RisaRepository==0.1
Products.RisaSimilarityTool==0.1
Products.RisaSite==0.1
Products.RisaWorkgroup==0.1
Products.SimpleAttachment==3.0.1-rhaptosdev-r30378
Products.Siyavula==0.2
Products.UniFile==0.2
Products.ZAnnot==0.5
Products.ZPsycopgDA==2.4.6-rhaptos
Products.Zope-Hotfix-CVE-2010-3198==1.0
argparse==1.2.1
demjson==1.6
egenix-mx-base==3.2.5
elementtree==1.2.7-20070827-preview
lxml==3.3.6
plone.recipe.zope2instance==3.6
psycopg2==2.4.6
python-memcached==1.48
rhaptos.atompub.plone==0.2
rhaptos.cnxmlutils==1.2
rhaptos.mathjax==1.1
rhaptos.swordservice.plone==1.1
simplejson==1.9.2
zc.buildout==1.4.4
zc.recipe.egg==1.2.2

===============================

```

We might want to include `update_versions.yml` in the other playbooks like `archive.yml`, `frontend.yml` etc?

### CODE REVIEW

File: update_versions.yml [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+    - shell: /var/cnx/venvs/publishing/bin/pip freeze
+      register: publishing_full_versions
+      delegate_to: "{{ item }}"
+      with_items:
+        - "{{ groups.publishing|first }}"
+
+- name: get legacy versions
+  hosts: frontend
+  tasks:
+    - shell: echo 'import pip; pip.main(["freeze"])' | /var/lib/cnx/cnx-buildout/bin/instance debug
```
Comment (pumazi) [inline update_versions.yml]: Nice workaround 🎈

**Comment by kerwinso** [comment]
@karenc This is great, but a little difficult to scan if I'm checking cnx package versions. I noticed that https://qa.cnx.org/version.txt looks different than the snippet in your [original comment](https://github.com/Connexions/cnx-deploy/pull/397#issue-264616254).

Instead of: 
`-e git+https://github.com/Connexions/cnx-db.git@59cc727d29275d50d2ba4ccfab4e9e1deda4f8ff#egg=cnx_db` 

can it be formatted to something like:

`cnx-db==59cc727d29275d50d2ba4ccfab4e9e1deda4f8ff`
? Or do the devs need the rest of that info? (I also noticed this formatting on the line for `cssselect2`.)

Also, how will tags be handled? My assumption is this on staging, the cnx versions will only be tags (due to our policy of only deploying tagged code to staging and prod), but on dev and qa environments we can expect to see a mix of tags and shas.

**Comment by karenc** [comment]
@kerwinso Yeah I can format it for `cnx-archive` and `cnx-publishing`, the rest I think will still have the github url so that devs can copy and paste that to install the exact versions locally.

The example version.txt in the pull request description is how it'll look like for staging and production that uses uploaded pypi packages instead of github.  (so `cnx-archive==2.6.1`)

**Comment by karenc** [comment]
Oh but perhaps we can separate the `version.txt` which might show just the packages we care about (packages that contain `cnx`?) and have the full versions in `full_version.txt` (which developers might find more useful)?

**Comment by kerwinso** [comment]
@karenc I'm in favor of just having one definitive `version.txt` file, reduces confusion.

**Comment by philschatz** [comment]
Ooh, there's been discussion around this today in `#devops` and `#manifestly`: https://openstax.slack.com/archives/C0MRT16UD/p1507840037000087

In there, I suggested having 2 files: `/rev.txt` (or something more descriptive) and `/history.log` . These address the following stories:

1. I want to know which versions of our code (not 3rd-party-packages) is running in an environment
1. I want to know which version of code was running last month

It seems like these files should include the frontend and backend repositories and my preference would be to include the repositories of all services in the environment, not just the server (or have a separate file for just the server info). For example, `publishing.cnx.org/rev.txt` would be identical to `cnx.org/rev.txt` as well as `archive.cnx.org/rev.txt`.

It would be nice (& maybe a requirement?) to have a way to quickly compare these `/rev.txt` files across environments to confidently say "production is running the same code that was tested on qa and staging". (A sha hash of the text file would suffice; manifestly was originally designed so that the commit sha would do this job)


---

Here are some stories that came up:

Some examples of “wants”:
1. As ops I don’t want to **require** GitHub when deploying to production so I can deploy or roll back at any time
1. As QA I want to be confident that the code on production is the same code that I tested on QA
1. As QA I want to minimize the number of typos and copy/pasta errors that can occur during deployment
1. As a dev I want to know what code is running on any environment
1. As a dev/QA/CS I want to know what was running on production last month so I can fix a bug
1. As a researcher I want to rebuild what was running on production a year ago

**Comment by karenc** [comment]
> 2. As QA I want to be confident that the code on production is the same code that I tested on QA

This is not going to be possible.  Code always get updated on QA, production runs pinned 3rd party libraries.  There is only one server for QA, there are 11 for production.  There are just too many differences between the two.  I think a more reasonable story would be "the code on production is the same as staging".

**Comment by karenc** [comment]
@philschatz There is no need to block this just because there are things you want to possibly change in the future.  We can always change it later.  If you have any suggestions for changes now, I can try to fix up the PR.

**Comment by philschatz** [comment]
ok, makes sense 😄 . In that case I have a few questions:

1. It seems that this format is specific to python packages, can `python` be in the name of the file?
1. It is difficult to quickly see which versions of **our** packages are installed. I see that `webview`, `archive`, and `publishing` are at the top which is nice but:
    1. will all the staging hosts have duplicate versions in this txt file? (I think that would be unnecessary)
    1. it is nice to see the pinned versions of our other packages... could they be sorted to be at the top?
1. can this file be JSON so that other services can easily pull this information? This would make it easier to write a Dashboard for QA that shows what is running on every environment

**Comment by karenc** [comment]
1. I think we can have a `python_version.txt` or something like that which doesn't have webview.
2. i. No, this `version.txf` is only on the servers with webview installed (so nginx can serve the file)
   ii. I think we can get a list of pinned packages from [archive-requirements.txt](https://github.com/Connexions/cnx-deploy/blob/master/environments/prod/files/archive-requirements.txt) which has the packages for what should be pinned, rather than what's on the server... :thinking:
3. I think we should add a `version.json` for the json version.  Json isn't naturally easy to read for humans, but I agree there should be a json version.

**Comment by pumazi** [comment]
@karenc, in a followup PR, please add the version of cnx-deploy to this file. Seems only right to include this since everything else is included. :kitchen_sink:

**Comment by karenc** [comment]
This creates `version.txt`, `python-version.txt` and `history.txt` in
`/var/www/webview` on all the frontend servers.

`version.txt` has the versions of webview, cnx-archive, cnx-publishing
and cnx-deploy in json format:

```json
{"date": "2017-10-19 09:20:58 PDT",
 "webview": "3c169baa52a6c5e82b5d7d0dd8a0e20c12c57f93 (v0.23.1-7-g3c169ba)",
 "cnx-archive": "ff577015a0a66d33756c8e1d68e2dba68a600ac1",
 "cnx-publishing": "5751cb983b14a40d0ca642ac884d122d9d1248d5",
 "cnx-deploy": "v2.2.4-44-gf9ebee3"
}
```

`python-version.txt` has the versions of everything in pip for cnx-archive,
cnx-publishing and zope in (kind of) python requirements.txt format:

```
# 2017-10-19 09:20:59 PDT
# Archive full versions from local.cnx.org:
certifi==2017.4.17
chardet==3.0.4
-e git+https://github.com/Connexions/cnx-archive.git@ff577015a0a66d33756c8e1d68e2dba68a600ac1#egg=cnx_archive
-e git+https://github.com/Connexions/cnx-db.git@bc6cb3ce5cbf9a302e2f7b8f49fe5f28c8606d00#egg=cnx_db
-e git+https://github.com/Connexions/cnx-epub.git@84e03d39958ae35f25c3bded8579eb9819c1b68f#egg=cnx_epub
-e git+https://github.com/Connexions/cnx-query-grammar.git@8484d2caed62aa5cdf1e0a631b1301a73cd5dbe8#egg=cnx_query_grammar
coverage==4.4.1
db-migrator==1.0.2
hupper==1.0
idna==2.5
Jinja2==2.9.6
lxml==3.8.0
MarkupSafe==1.0
mock==1.0.1
parsimonious==0.7.0
PasteDeploy==1.5.2
pkg-resources==0.0.0
plaster==0.5
plaster-pastedeploy==0.4.1
psycopg2==2.7.3
py==1.4.34
pyramid==1.9.1
pyramid-jinja2==2.7
pyramid-sawing==1.1.3
pytest==3.2.0
pytest-cov==2.5.1
python-memcached==1.58
pytz==2017.2
PyYAML==3.12
repoze.lru==0.6
requests==2.18.2
-e git+https://github.com/Connexions/rhaptos.cnxmlutils.git@23e0c4ddcded627c4b42dcec6fc90a2efaa37cfb#egg=rhaptos.cnxmlutils
six==1.10.0
translationstring==1.3
tzlocal==1.4
urllib3==1.22
venusian==1.1.0
waitress==1.0.2
WebOb==1.7.3
zope.deprecation==4.2.0
zope.interface==4.4.2

# Publishing full versions from local.cnx.org:
amqp==2.2.1
Beaker==1.9.0
beautifulsoup4==4.6.0
billiard==3.5.0.3
celery==4.1.0
certifi==2017.4.17
chardet==3.0.4
-e git+https://github.com/Connexions/cnx-archive.git@ff577015a0a66d33756c8e1d68e2dba68a600ac1#egg=cnx_archive
-e git+https://github.com/Connexions/cnx-db.git@bc6cb3ce5cbf9a302e2f7b8f49fe5f28c8606d00#egg=cnx_db
-e git+https://github.com/Connexions/cnx-easybake.git@c109c756c97cdfc90d2bcca8dee173922662f28e#egg=cnx_easybake
-e git+https://github.com/Connexions/cnx-epub.git@84e03d39958ae35f25c3bded8579eb9819c1b68f#egg=cnx_epub
-e git+https://github.com/Connexions/cnx-publishing.git@5751cb983b14a40d0ca642ac884d122d9d1248d5#egg=cnx_publishing
-e git+https://github.com/Connexions/cnx-query-grammar.git@8484d2caed62aa5cdf1e0a631b1301a73cd5dbe8#egg=cnx_query_grammar
-e git+https://github.com/Connexions/cnx-recipes.git@1dc9cdcd7e94e11b3cd8489056359bd74d77ccb8#egg=cnx_recipes
configparser==3.5.0
coverage==4.4.1
cssselect==1.0.1
-e git+https://github.com/Connexions/cssselect2.git@560b4c432fcea05e4694ad6367b5a0e4d9fe2d33#egg=cssselect2
db-migrator==1.0.2
doc8==0.8.0
docopt==0.6.2
docutils==0.14
enum34==1.1.6
faulthandler==2.6
flake8==3.4.1
funcsigs==1.0.2
gitdb2==2.0.2
GitPython==2.1.5
hupper==1.0
idna==2.5
Jinja2==2.9.6
kombu==4.1.0
lxml==3.8.0
MarkupSafe==1.0
mccabe==0.6.1
mock==1.0.1
-e git+https://github.com/Connexions/openstax-accounts.git@f96049d99ba3abb3c482508c0c19d87b3cecb005#egg=openstax_accounts
parsimonious==0.7.0
PasteDeploy==1.5.2
pbr==3.1.1
pep8==1.7.0
pika==0.11.0
pkg-resources==0.0.0
plaster==0.5
plaster-pastedeploy==0.4.1
psycopg2==2.7.3.1
py==1.4.34
pycodestyle==2.3.1
pyflakes==1.5.0
pyramid==1.9.1
pyramid-jinja2==2.7
pyramid-multiauth==0.9.0
pyramid-sawing==1.1.3
pytest==3.2.0
pytest-cov==2.5.1
pytest-faulthandler==1.3.1
pytest-mock==1.6.2
pytest-runner==2.11.1
pytest-timeout==1.2.0
python-memcached==1.58
pytz==2017.2
PyYAML==3.12
repoze.lru==0.6
requests==2.18.2
restructuredtext-lint==1.1.1
-e git+https://github.com/Connexions/rhaptos.cnxmlutils.git@23e0c4ddcded627c4b42dcec6fc90a2efaa37cfb#egg=rhaptos.cnxmlutils
sanction==0.4.1
six==1.10.0
smmap2==2.0.3
SQLAlchemy==1.1.12
stevedore==1.26.0
tinycss2==0.5
translationstring==1.3
tzlocal==1.4
urllib3==1.22
venusian==1.1.0
vine==1.1.4
waitress==1.0.2
webencodings==0.5.1
WebOb==1.7.3
WebTest==2.0.28
zope.deprecation==4.2.0
zope.interface==4.4.2

# Legacy full versions from local.cnx.org:
Jinja2==2.6
Pillow==1.7.6
## FIXME: could not find svn URL in dependency_links for this package:
Products.AdvancedQuery==2.2-rhaptosdev-r30378
## FIXME: could not find svn URL in dependency_links for this package:
Products.CMFDiffTool==0.3.3dev-r30461
Products.CMFSquidTool==1.5.1
Products.CNXCaching==0.2
Products.CNXContent==0.47
Products.CNXMLDocument==0.49
Products.CNXMLTransforms==0.23
Products.CNXPloneSite==0.76.2dev
Products.CacheSetup==1.2.1
Products.CatalogMemberDataTool==1.5
Products.ClockServer==0.2-Zope2.9dev
Products.ExtZSQL==0.9
## FIXME: could not find svn URL in dependency_links for this package:
Products.ExternalFile==0.1-rhaptosdev-r30470
Products.ExternalStorage==0.8.1
Products.FSImportTool==0.8
Products.FeatureArticle==0.10.1
## FIXME: could not find svn URL in dependency_links for this package:
Products.Five==1.4.5-rhaptosdev-r32037
Products.LensOrganizer==0.3.1
Products.Lensmaker==0.25
Products.Lineup==0.6
Products.LinkMapTool==0.9
## FIXME: could not find svn URL in dependency_links for this package:
Products.LocalFS==1.7-rhaptosdev-r30378
## FIXME: could not find svn URL in dependency_links for this package:
Products.ManagableIndex==1.7.3-rhaptosdev-r30378
## FIXME: could not find svn URL in dependency_links for this package:
Products.MasterSelectWidget==0.2.3-rhaptosdev-r30378
Products.MathEditor==0.4.1
## FIXME: could not find svn URL in dependency_links for this package:
Products.NoHeaderFieldContinuation==0.1-rhaptosdev-r30378
## FIXME: could not find svn URL in dependency_links for this package:
Products.OFolder==1.0-rhaptosdev-r30378
Products.PageCacheManager==1.2
Products.PloneHotfix20110531==2.0
Products.PloneHotfix20110720==1.2
Products.PloneHotfix20121106==1.2
Products.PloneHotfix20130618==1.1
Products.PolicyHTTPCacheManager==1.2
## FIXME: could not find svn URL in dependency_links for this package:
Products.References==0.10-rhaptosdev-r30378
Products.RhaptosBugTrackingTool==0.1.3
Products.RhaptosCacheTool==0.2.1
Products.RhaptosCatFixes==0.2
Products.RhaptosCollaborationTool==0.10.1
Products.RhaptosCollection==0.77
Products.RhaptosContent==0.66
Products.RhaptosHitCountTool==0.12
Products.RhaptosModuleEditor==0.86
Products.RhaptosModuleStorage==1.1.3
Products.RhaptosPDFLatexTool==0.15.1dev
Products.RhaptosPatchTool==0.7.2
Products.RhaptosPrint==0.104
Products.RhaptosRepository==1.4.2
Products.RhaptosSimilarityTool==0.13
Products.RhaptosSite==1.46
Products.RhaptosSword==2.3.2dev
Products.RhaptosTest==1.0
Products.RhaptosWorkgroup==0.27
Products.RisaCollection==0.1
Products.RisaHitCountTool==0.1
Products.RisaModuleEditor==0.1
Products.RisaModuleStorage==0.1
Products.RisaPDFLatexTool==0.1
Products.RisaPatchTool==0.1
Products.RisaRepository==0.1
Products.RisaSimilarityTool==0.1
Products.RisaSite==0.1
Products.RisaWorkgroup==0.1
## FIXME: could not find svn URL in dependency_links for this package:
Products.SimpleAttachment==3.0.1-rhaptosdev-r30378
Products.Siyavula==0.2
Products.UniFile==0.2
Products.ZAnnot==0.5
Products.ZPsycopgDA==2.4.6-rhaptos
Products.Zope-Hotfix-CVE-2010-3198==1.0
argparse==1.2.1
db-migrator==1.0.2
demjson==1.6
egenix-mx-base==3.2.5
elementtree==1.2.7-20070827-preview
lxml==3.3.6
plone.recipe.zope2instance==3.6
psycopg2==2.4.6
python-memcached==1.48
rhaptos.atompub.plone==0.2
rhaptos.cnxmlutils==1.2
rhaptos.mathjax==1.1
rhaptos.swordservice.plone==1.1
simplejson==1.9.2
zc.buildout==1.4.4
zc.recipe.egg==1.2.2
```

`history.txt` prepends the output of `version.txt` and `python-version.txt` to
the existing `history.txt` so we get a list of package versions with dates.

Close #407

**Comment by karenc** [comment]
If everyone is ok with this pull request, let's merge it and get version.txt out in the next deploy.

**Comment by dtwilliamson** [comment]
I see a reference to "local.cnx.org" in the package listing. This server doesn't exist.

**Comment by karenc** [comment]
@dtwilliamson `local.cnx.org` is the domain name for the vm environment which I use locally to generate the example files.

File: update_versions.yml [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+           "webview": "{{ webview_version.stdout }}",
+           "cnx-archive": "{{ archive_version.results.0.stdout|regex_replace('^[^@]*@([^#]*).*$', '\1')|regex_replace('^[^=]*==') }}",
+           "cnx-publishing": "{{ publishing_version.results.0.stdout|regex_replace('^[^@]*@([^#]*).*$', '\1')|regex_replace('^[^=]*==') }}",
+           "cnx-deploy": "{{ cnx_deploy_version.stdout }}"
+          }
+          EOF
+      args:
+        executable: /bin/sh
+    - name: purge cached version.txt
+      command: "curl -X PURGE 'http://{{ frontend_domain }}:{{ varnish_port }}/version.txt'"
```
Comment (reedstrm) [inline update_versions.yml]: If I read correctly, the varnish config is currently setup to only accept PURGE requests from localhost. Will this work?

Comment (karenc) [reply update_versions.yml]: Oh... may be not... I'll have to check

Comment (karenc) [reply update_versions.yml]: You're right, it can't be done.  I'll update the PR

Comment (reedstrm) [reply update_versions.yml]: Any chance we could do that the TODO in the config says, and template in the IPs of the machine where this is running? (we're maybe going to need that for re-baking, eventually anyway)

Comment (karenc) [reply update_versions.yml]: Oh actually the task is executed on the `frontend` hosts and they are the ones with varnish installed. :relieved: so it'll work.

Comment (karenc) [reply update_versions.yml]: @reedstrm I'm not sure what you mean

Comment (reedstrm) [reply update_versions.yml]: https://github.com/Connexions/cnx-deploy/blob/master/roles/varnish/templates/etc/varnish/varnish.vcl#L110
says "# TODO Include IP addresses of other serving hosts."
This is the ACL. Note that it explicitly defines 127.0.0.1, which is not the network frontend_domain is in. Even though localhost ans 127.0.0.1 are in the acl for PURGE, a request that is not frontend_domain end up catching the hard return of 750. I'll add a oneliner patch.

Comment (reedstrm) [reply update_versions.yml]: The alternative would be to add branch recognizing incoming from localhost, and allow  that through, then change the PURGE above to `http://localhost`

Comment (karenc) [reply update_versions.yml]: I have `local.cnx.org` pointing to `127.0.0.1`:

```
karen@xenial-openstax:/var/cnx/venvs$ cat /etc/hosts
127.0.0.1       localhost xenial-openstax.local local.cnx.org archive.local.cnx.org authoring.local.cnx.org legacy.local.cnx.org zope.local.cnx.org
```

So it works for me locally:

```
karen@xenial-openstax:/var/cnx/venvs$ curl -X PURGE 'http://local.cnx.org:8990/version.txt'
<!DOCTYPE html>
<html>
  <head>
    <title>200 Ban added</title>
  </head>
  <body>
    <h1>Error 200 Ban added</h1>
    <p>Ban added</p>
    <h3>Guru Meditation:</h3>
    <p>XID: 685955</p>
    <hr>
    <p>Varnish cache server</p>
  </body>
</html>
```

`localhost` doesn't seem to work though:

```
karen@xenial-openstax:/var/cnx/venvs$ curl -X PURGE 'http://localhost:8990/version.txt'
<!DOCTYPE html>
<html>
  <head>
    <title>750 Moved Permanently</title>
  </head>
  <body>
    <h1>Error 750 Moved Permanently</h1>
    <p>Moved Permanently</p>
    <h3>Guru Meditation:</h3>
    <p>XID: 686338</p>
    <hr>
    <p>Varnish cache server</p>
  </body>
</html>
```

Anyway, with your commit, even if `local.cnx.org` doesn't point to `127.0.0.1`, it works so :+1:

**Comment by kerwinso** [comment]
Related to this Trello card: https://trello.com/c/gKqLyiS3

**Pull request merged**